workspaces/github-issues/plugins/search-backend-module-github-issues/src/collators/GithubIssuesCollatorFactory.ts

```diff
@@ -0,0 +1,234 @@
+import { Readable } from 'stream';
+
+/**
+ * Extended IndexableDocument with GitHub specific properties
+ *
+ * @public
+ */
+export interface GithubIssuesDocument extends IndexableDocument {
+  title: string;
+  createdAt: string;
+  url: string;
+  repository: {
+    name: string;
+  };
+  author: string;
+  labels: string[];
+}
+
+export type GithubIssuesCollatorFactoryOptions = {
+  logger: LoggerService;
+  config: RootConfigService;
+  githubCredentialsProvider: GithubCredentialsProvider;
+  integrations: ScmIntegrationRegistry;
+};
+
+type GraphQL = typeof graphql;
+
+/**
+ * Search collator responsible for collecting GitHub issues and discussions to index.
+ *
+ * @public
+ */
+export class GithubIssuesCollatorFactory implements DocumentCollatorFactory {
+  public readonly type: string = 'github';
+
+  private readonly logger: LoggerService;
+  private readonly config: Config;
+
+  private readonly integrations: ScmIntegrationRegistry;
+  private readonly githubCredentialsProvider: GithubCredentialsProvider;
+
+  constructor(options: GithubIssuesCollatorFactoryOptions) {
+    this.integrations = options.integrations;
+    this.githubCredentialsProvider = options.githubCredentialsProvider;
+    this.logger = options.logger;
+    this.config = options.config;
+  }
+
+  static fromConfig(
+    config: Config,
+    options: {
+      logger: LoggerService;
+      githubCredentialsProvider?: GithubCredentialsProvider;
+    },
+  ) {
+    const integrations = ScmIntegrations.fromConfig(config);
+
+    return new GithubIssuesCollatorFactory({
+      config: config,
+      logger: options.logger,
+      githubCredentialsProvider:
+        options.githubCredentialsProvider ||
+        DefaultGithubCredentialsProvider.fromIntegrations(integrations),
+      integrations,
+    });
+  }
+
+  async getCollator() {
+    return Readable.from(this.execute());
+  }
+
+  async *execute(): AsyncGenerator<GithubIssuesDocument> {
+    this.logger.info('Starting collation of GitHub issues.');
+
+    const documents = await this.fetchGithubIssuesDocuments();
+
+    for (const document of documents) {
+      if (!document.title || !document.url) {
+        continue;
+      }
+      if (!document.author || !document.repository) {
+        this.logger.warn('Document missing author or repository information');
+        continue;
+      }
+      const labels =
+        document.labels?.nodes
+          ?.map(label => label?.name)
+          .filter(
+            (name): name is string => name !== undefined && name !== null,
+          ) ?? [];
+
+      yield {
+        ...document,
+        text: document.title,
+        location: document.url,
+        author: document.author.login,
+        labels,
+      };
+    }
+
+    this.logger.info('Finished collation of GitHub issues.');
+  }
+
+  private async fetchGithubIssuesDocuments(): Promise<Issue[] | Discussion[]> {
+    const { client } = await this.createClient();
```

Comment:
we could move `createClient` to the parent function, so that we instantiate the client only once

Reply:
Makes sense 🫡